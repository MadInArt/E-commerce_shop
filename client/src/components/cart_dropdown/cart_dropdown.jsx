import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { createStructuredSelector } from "reselect";

import { selectCartItems } from "../../redux/cart/cart-selector";
import CartItem from "../cart-item/cart-item";
import { setToggleCart } from "../../redux/cart/cart-action";

import {
  CartDropdownContainer,
  CartDropdownButton,
  EmptyMessageContainer,
  CartItemsContainer,
} from "./cart_dropdown.styles";

export const CartDropdown = ({ cartItems, history, dispatch }) => (
  <CartDropdownContainer>
    <CartItemsContainer>
      {cartItems.length ? (
        cartItems.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))
      ) : (
        <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
      )}
    </CartItemsContainer>
    <CartDropdownButton
      onClick={() => {
        history.push("/checkout");
        dispatch(setToggleCart());
      }}
    >
      GO TO PAYMENT
    </CartDropdownButton>
  </CartDropdownContainer>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});
export default React.memo(withRouter(connect(mapStateToProps)(CartDropdown)));
