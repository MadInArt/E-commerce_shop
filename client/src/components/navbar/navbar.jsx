import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import { signOutStart } from "../../redux/user/user-action";
import { selectCartHidden } from "../../redux/cart/cart-selector";
import { selectCurrentUser } from "../../redux/user/user-selector";
import { ReactComponent as Logo } from "../../assets/logo.svg";
import CartItem from "../cart-icon/cart-icon";
import CartDropdown from "../cart_dropdown/cart_dropdown";

import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink,
} from "./navbar.styles";

export const Navbar = ({ currentUser, hidden, signOutStart }) => (
  <HeaderContainer>
    <LogoContainer to="/">
      <Logo className="logo" />
    </LogoContainer>
    <OptionsContainer>
      <OptionLink to="/">HOME</OptionLink>
      <OptionLink to="/shop">SHOP</OptionLink>
      <OptionLink to="/contact">CONTACT</OptionLink>
      {currentUser ? (
        <div className="links" onClick={signOutStart}>
          SIGN OUT
        </div>
      ) : (
        <OptionLink to="/signin">SIGN IN</OptionLink>
      )}
      <CartItem />
    </OptionsContainer>
    {hidden ? null : <CartDropdown />}
  </HeaderContainer>
);

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden,
});
const mapDispatchToProps = (dispatch) => ({
  signOutStart: () => dispatch(signOutStart()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Navbar);
