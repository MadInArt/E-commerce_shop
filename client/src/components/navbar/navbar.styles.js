import styled from "styled-components";
import { Link } from "react-router-dom";

export const NavbarContainer = styled.div`
  height: 70px;
  width: 100%;
  display: flex;
  justify-content: space-between;
  margin-bottom: 25px;
`;
export const NavbarLinks = styled.div`
  width: 50%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;

  a {
    padding: 10px 15px;
    cursor: pointer;
  }
`;

export const NavbarLogoContainer = styled(Link)`
  height: 100%;
  width: 70px;
  padding: 25px;
`;
