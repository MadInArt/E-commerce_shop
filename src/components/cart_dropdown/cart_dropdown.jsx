import React from "react";
import { connect } from "react-redux";

import "./cart_dropdown.scss";
import CustomButton from "../button/button";
import CartItem from '../cart-item/cart-item';

const CartDropdown = ({ cartItems }) => (
  <div className="dropdown">
    <div className="items">
      {cartItems.map(cartItem => (
        <CartItem key={cartItem.id} item={cartItem} />
      ))}
    </div>
    <CustomButton className="button">GO TO PAYMENT</CustomButton>
  </div>
);

const mapStateToProps = ({ cart: { cartItems } }) => ({
  cartItems,
});
export default connect(mapStateToProps)(CartDropdown);
