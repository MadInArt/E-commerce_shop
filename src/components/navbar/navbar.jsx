import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";

import { ReactComponent as Logo } from "../../assets/logo.svg";
import CartItem from "../cart-icon/cart-icon";
import CartDropdown from "../cart_dropdown/cart_dropdown";
import { auth } from "../../firebase/firebase.utils";

import "./navbar.scss";

const Navbar = ({ currentUser, hidden }) => (
  <div className="header">
    <Link to="/" className="logo-container">
      <Logo className="logo" />
    </Link>
    <div className="links">
      <Link to="/shop">SHOP</Link>
      <Link to="/contact">CONTACT</Link>
      {currentUser ? (
        <div className="links" onClick={() => auth.signOut()}>
          SIGN OUT
        </div>
      ) : (
        <Link className="links" to="/signin">
          SIGN IN
        </Link>
      )}
      <CartItem/>
    </div>
    {hidden ? null : <CartDropdown />}
  </div>
);

const mapStateToProps = ({ user: { currentUser }, cart: { hidden } }) => ({
  currentUser,
  hidden,
});

export default connect(mapStateToProps)(Navbar);
