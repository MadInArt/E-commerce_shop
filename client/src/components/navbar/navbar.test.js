import React from "react";
import { shallow } from "enzyme";

import { Navbar } from "./navbar";
import { CartDropdown } from "../cart_dropdown/cart_dropdown";

describe("Navbar component tests", () => {
  let wrapper;
  let mockSignOutStart;

  beforeEach(() => {
    mockSignOutStart = jest.fn();

    const mockProps = {
      hidden: true,
      currentUser: {
        uid: "123",
      },
      signOutStart: mockSignOutStart,
    };

    wrapper = shallow(<Navbar {...mockProps} />);
  });

  test("should render Navbar component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("if currentUser is null", () => {
    test("should render sign in link", () => {
      const mockProps = {
        hidden: true,
        currentUser: null,
        signOutStart: mockSignOutStart,
      };

      const newWrapper = shallow(<Navbar {...mockProps} />);

      expect(newWrapper.find("OptionLink").at(2).text()).toBe("SIGN IN");
    });
  });

  describe("if hidden is true", () => {
    test("should not render CartDropdown", () => {
      expect(wrapper.exists(CartDropdown)).toBe(false);
    });
  });
});
