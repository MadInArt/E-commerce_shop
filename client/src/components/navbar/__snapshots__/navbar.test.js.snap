// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component tests should render Navbar component 1`] = `
<styled.div>
  <Styled(Link)
    to="/"
  >
    <ForwardRef(SvgLogo)
      className="logo"
    />
  </Styled(Link)>
  <styled.div>
    <OptionLink
      to="/shop"
    >
      SHOP
    </OptionLink>
    <OptionLink
      to="/contact"
    >
      CONTACT
    </OptionLink>
    <div
      className="links"
      onClick={[MockFunction]}
    >
      SIGN OUT
    </div>
    <Connect(CartIcon) />
  </styled.div>
</styled.div>
`;
